`P2286R8_text_formatting_tuple` fails on non-English locale

```
FAIL: std :: tests/P2286R8_text_formatting_tuple:22 (1448 of 1578)
******************** TEST 'std :: tests/P2286R8_text_formatting_tuple:22' FAILED ********************
Build setup steps:
Build steps:
Build step failed unexpectedly.
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64\cl.EXE" "C:\Project\STL\tests\std\tests\P2286R8_text_formatting_tuple\test.cpp" "-IC:\Project\STL\out\x64\out\inc" "-IC:\Project\STL\llvm-project\libcxx\test\support" "-IC:\Project\STL\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/w14668" "/w15267" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/w14640" "/Zc:threadSafeInit-" "/EHsc" "/std:c++latest" "/MTd" "/permissive" "-fsanitize=address" "/Zi" "-FeC:\Project\STL\out\x64\tests\std\tests\P2286R8_text_formatting_tuple\Output\22\P2286R8_text_formatting_tuple.exe" "-link" "-LIBPATH:C:\Project\STL\out\x64\out\lib\amd64" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.40.33807\lib\x64" "/MANIFEST:EMBED" "/debug"
Exit Code: 2 (0x2)
Standard Output:
--
test.cpp
C:\Project\STL\tests\std\tests\P2286R8_text_formatting_tuple\test.cpp(161): error C2220: the following warning is treated as an error
C:\Project\STL\tests\std\tests\P2286R8_text_formatting_tuple\test.cpp(161): warning C4566: character represented by universal-character-name '\u00D6' cannot be represented in the current code page (1251)
C:\Project\STL\tests\std\tests\P2286R8_text_formatting_tuple\test.cpp(162): warning C4566: character represented by universal-character-name '\u00D6' cannot be represented in the current code page (1251)
--
```

--- a/stl/inc/format
+++ b/stl/inc/format
@@ -3961,6 +3961,13 @@ _NODISCARD size_t formatted_size(const locale& _Loc, const wformat_string<_Types
 _FMT_P2286_END
 
 #if _HAS_CXX23
+template <class _CharT>
+_NODISCARD int _Measure_display_width(const basic_string_view<_CharT> _Value) {
+    int _Width = -1;
+    (void) _Measure_string_prefix(_Value, _Width);
+    return _Width;
+}
+
 enum class _Fmt_tuple_type : uint8_t { _None, _Key_value, _No_brackets };
 
 template <class _CharT>
@@ -4127,8 +4134,9 @@ protected:
         }(index_sequence_for<_ArgTypes...>{});
         _STD _Copy_unchecked(_Closing_bracket._Unchecked_begin(), _Closing_bracket._Unchecked_end(), _Tmp_ctx.out());
 
-        return _STD _Write_aligned(_Fmt_ctx.out(), static_cast<int>(_Tmp_buf.size()), _Format_specs, _Fmt_align::_Left,
-            [&](typename _FormatContext::iterator _Out) {
+        const int _Width = _Measure_display_width<_CharT>(_Tmp_buf);
+        return _STD _Write_aligned(
+            _Fmt_ctx.out(), _Width, _Format_specs, _Fmt_align::_Left, [&](typename _FormatContext::iterator _Out) {
                 return _STD _Fmt_write(_STD move(_Out), basic_string_view<_CharT>{_Tmp_buf});
             });
     }

--- a/tests/std/tests/P2286R8_text_formatting_escaping_legacy_text_encoding/test.cpp
+++ b/tests/std/tests/P2286R8_text_formatting_escaping_legacy_text_encoding/test.cpp
@@ -17,6 +17,15 @@ void test_escaped_string() {
     assert(format("{:?}", "\x81\x40\x40\x81") == "\"\\u{3000}\x40\\x{81}\"");
 }
 
+template <class TupleOrPair>
+void test_tuple_or_pair_escaping(TupleOrPair&& input) {
+    get<1>(input) = "hell\uff2f"; // U+FF2F FULLWIDTH LATIN CAPITAL LETTER O
+    assert(format("{}", input) == "('*', \"hell\uff2f\")");
+    assert(format("{:#^16}", input) == "('*', \"hell\uff2f\")#");
+}
+
 int main() {
     test_escaped_string();
+    test_tuple_or_pair_escaping(make_pair('*', ""));
+    test_tuple_or_pair_escaping(make_tuple('*', ""));
 }

--- a/tests/std/tests/P2286R8_text_formatting_escaping_utf8/test.cpp
+++ b/tests/std/tests/P2286R8_text_formatting_escaping_utf8/test.cpp
@@ -21,15 +21,32 @@ void test_escaped_string() {
     assert(format("{:?}", "a\u0300") == "\"a\u0300\"");
 }
 
-int main() {
+template <class TupleOrPair>
+void test_tuple_or_pair_escaping(TupleOrPair&& input) {
+    get<1>(input) = "hell\u00d6"; // U+00D6 LATIN CAPITAL LETTER O WITH DIAERESIS
+    assert(format("{}", input) == "('*', \"hell\u00d6\")");
+    assert(format("{:#^16}", input) == "#('*', \"hell\u00d6\")#");
+
+    get<1>(input) = "hell\uff2f"; // U+FF2F FULLWIDTH LATIN CAPITAL LETTER O
+    assert(format("{}", input) == "('*', \"hell\uff2f\")");
+    assert(format("{:#^16}", input) == "('*', \"hell\uff2f\")#");
+}
+
+void run_test() {
     test_escaped_string();
+    test_tuple_or_pair_escaping(make_pair('*', ""));
+    test_tuple_or_pair_escaping(make_tuple('*', ""));
+}
+
+int main() {
+    run_test();
 
     assert(setlocale(LC_ALL, ".1252") != nullptr);
-    test_escaped_string();
+    run_test();
 
     assert(setlocale(LC_ALL, ".932") != nullptr);
-    test_escaped_string();
+    run_test();
 
     assert(setlocale(LC_ALL, ".UTF-8") != nullptr);
-    test_escaped_string();
+    run_test();
 }

--- a/tests/std/tests/P2286R8_text_formatting_tuple/test.cpp
+++ b/tests/std/tests/P2286R8_text_formatting_tuple/test.cpp
@@ -157,9 +157,16 @@ void test_escaping(TestFunction check, TupleOrPair&& input) {
     check(SV(R"(('\u{0}', ""))"), SV("{}"), input);
 
     // String
-    get<0>(input) = CharT('*');
-    get<1>(input) = SV("hell\u00d6");
-    check(SV("('*', \"hell\u00d6\")"), SV("{}"), input);
+    if constexpr (is_same_v<CharT, wchar_t>) {
+        get<0>(input) = L'*';
+        get<1>(input) = L"hell\u00d6"; // U+00D6 LATIN CAPITAL LETTER O WITH DIAERESIS
+        check(L"('*', \"hell\u00d6\")"sv, L"{}"sv, input);
+        check(L"#('*', \"hell\u00d6\")#"sv, L"{:#^16}"sv, input);
+
+        get<1>(input) = L"hell\uff2f"; // U+FF2F FULLWIDTH LATIN CAPITAL LETTER O
+        check(L"('*', \"hell\uff2f\")"sv, L"{}"sv, input);
+        check(L"('*', \"hell\uff2f\")#"sv, L"{:#^16}"sv, input);
+    }
 }
 
 //

